Vectorize the numeric part of Earliest Aggregator

somu-imply opened the pull request:
Vectorizing the earliest aggregator
```Benchmark                        (query)  (rowsPerSegment)  (vectorize)  Mode  Cnt    Score    Error  Units
SqlExpressionBenchmark.querySql       42           5000000        false  avgt    5   86.661 ± 10.316  ms/op
SqlExpressionBenchmark.querySql       42           5000000        force  avgt    5   58.923 ±  9.156  ms/op
SqlExpressionBenchmark.querySql       43           5000000        false  avgt    5   76.705 ± 13.739  ms/op
SqlExpressionBenchmark.querySql       43           5000000        force  avgt    5   43.493 ±  7.034  ms/op
SqlExpressionBenchmark.querySql       44           5000000        false  avgt    5   75.060 ±  5.100  ms/op
SqlExpressionBenchmark.querySql       44           5000000        force  avgt    5   38.841 ±  1.883  ms/op
SqlExpressionBenchmark.querySql       45           5000000        false  avgt    5  189.588 ±  6.205  ms/op
SqlExpressionBenchmark.querySql       45           5000000        force  avgt    5   62.001 ±  9.902  ms/op
```

where
```
42 --> long
43 --> float
44 --> double
45 --> all three
```

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

TSFenwick commented on processing/src/main/java/org/apache/druid/query/aggregation/first/LongFirstVectorAggregator.java:
```diff
@@ -0,0 +1,66 @@
+ */
+public class LongFirstVectorAggregator extends NumericFirstVectorAggregator
+{
+  long firstValue;
```
private?

TSFenwick commented on processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java:
```diff
@@ -0,0 +1,173 @@
+    final boolean[] nullValueVector = valueSelector.getNullVector();
+    boolean nullAbsent = false;
+    firstTime = buf.getLong(position);
+    //check if nullVector is found or not
```
nit: spacing on comments are inconsistent

somu-imply replied:
adjusted

TSFenwick commented on processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java:
```diff
@@ -0,0 +1,173 @@
+  }
+
+  /**
+   *
```
would be good to have an overall description of the method  be in the javadocs.

somu-imply replied:
done

TSFenwick commented on processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java:
```diff
@@ -0,0 +1,173 @@
+  }
+
+  /**
+   *
```
would be good to have an overall description of the method  be in the javadocs.

somu-imply replied:
done

suneet-s commented on processing/src/main/java/org/apache/druid/query/aggregation/first/DoubleFirstVectorAggregator.java:
```diff
@@ -0,0 +1,66 @@
+
+/**
+ * Vectorized version of on heap 'last' aggregator for column selectors with type LONG..
+ */
```
stale javadocs. Similar comment for other `VectorAggregator` classes

somu-imply replied:
updated

suneet-s commented on sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:
```diff
@@ -995,6 +993,40 @@ public void testPrimitiveEarliestInSubquery() throws Exception
   }
 
+  @Test
+  public void testPrimitiveEarliestInSubqueryGroupBy() throws Exception
```
This doesn't look like a subquery that is being tested

somu-imply replied:
updated

suneet-s reviewed (commented):
```
|   0% (0/7)     |   0% (0/4)     |  73% (11/15)   | org/apache/druid/query/aggregation/first/FloatFirstAggregatorFactory.java
```

Overall LGTM. This is the only thing that slightly concerns me. Can you look into why there's no coverage for changes to this class, while there is coverage for the other aggregator factories.

somu-imply commented:
@suneet-s I have added additional tests that would cover the factory classes for all the three new additions. The coverage would improve now.
@TSFenwick I have updated as per your review comments.

suneet-s commented:
@somu-imply It looks like there are still some failing tests, can you fix them up please.

`TimeseriesQueryRunnerTest.testTimeseriesWithFirstLastAggregator` is one example

somu-imply commented:
Instead of conflict resolution that's gonna make things difficult, I'll make a fresh PR will close this once the new one's up and linked to this
